integration/osrmfiles/osrmtype/profile/properties.go

```diff
@@ -0,0 +1,133 @@
+// otherwise error will occrred.
+// Also, the parsing code only considers 64 bits machine, DOESN'T support 32 bits machine.
+type Properties struct {
+	TrafficSignalPenalty   int32 // NOTE that it's different from `traffic_light_penality`, actually it's been DEPRECATED already.
+	UTurnPenalty           int32
+	MaxSpeedForMapMatching float64
+
+	// 4 bool together takes 4 bytes
+	ContinueStraightAtWaypoint bool
+	UseTurnRestrictions        bool
+	LeftHandDriving            bool // DEPRECATED, see https://github.com/Telenav/osrm-backend/blob/8e094d1b90a31c090cdafa9854985343ea1e15c3/include/extractor/profile_properties.hpp#L129
+	FallbackToDuration         bool
+
+	WeightName        string
+	ClassNames        []string
+	ExcludableClasses []uint8
+
+	// uint + 2 bool together takes 8 bytes on 64 bits machine
+	WeightPrecision         uint
+	ForceSplitEdges         bool
+	CallTaglessNodeFunction bool
+}
+
+const (
+	propertiesBytes = 2344 // the C++ implementation layout actual size
+
+	// https://github.com/Telenav/osrm-backend/blob/8e094d1b90a31c090cdafa9854985343ea1e15c3/include/extractor/profile_properties.hpp#L132
+	weightNameBytes       = 256
+	classNameBytes        = 256
+	classNamesCount       = 8
+	excluableClassesCount = 8
+
+	lastPaddingBytes = 4 // last padding bytes due to C++ memory alignment
+)
+
+func (pr *Properties) Write(p []byte) (int, error) {
+	if len(p) != propertiesBytes {
+		return 0, fmt.Errorf("lua properties should always has %d bytes on 64-bit machine. Check whether anything changed on `struct ProfileProperties`, or memory alignment on machine", propertiesBytes)
+	}
+
+	var writeLen int
+	writeP := p
+
+	// first 16 bytes
+	pr.TrafficSignalPenalty = int32(binary.LittleEndian.Uint32(writeP))
+	pr.UTurnPenalty = int32(binary.LittleEndian.Uint32(writeP[4:]))
+	pr.MaxSpeedForMapMatching = math.Float64frombits(binary.LittleEndian.Uint64(writeP[8:]))
+	writeP = writeP[16:]
+	writeLen += 16
+
+	// 4 bool/4 bytes
+	if writeP[0] > 0 {
+		pr.ContinueStraightAtWaypoint = true
+	}
+	if writeP[1] > 0 {
+		pr.UseTurnRestrictions = true
+	}
+	if writeP[2] > 0 {
+		pr.LeftHandDriving = true
+	}
+	if writeP[3] > 0 {
+		pr.FallbackToDuration = true
+	}
+	writeP = writeP[4:]
+	writeLen += 4
+
+	// names
+	weightNameValidBytes := bytes.IndexByte(writeP[:weightNameBytes], 0)
+	if weightNameValidBytes > 0 {
+		pr.WeightName = string(writeP[:weightNameValidBytes])
+	} else {
+		pr.WeightName = string(writeP[:weightNameBytes])
+	}
+	writeP = writeP[weightNameBytes:]
+	writeLen += weightNameBytes
+	for i := 0; i < classNamesCount; i++ {
+		classNameValidBytes := bytes.IndexByte(writeP[:classNameBytes], 0)
+		if classNameValidBytes > 0 {
+			pr.ClassNames = append(pr.ClassNames, string(writeP[:classNameValidBytes]))
+		} else if classNameValidBytes == 0 {
+			pr.ClassNames = append(pr.ClassNames, "")
+		} else {
+			pr.ClassNames = append(pr.ClassNames, string(writeP[:classNameBytes]))
+		}
```

Comment:
1. Just confirm, class names like
```
0classa0classb0classc
```
2. when will be `0` in the head?

Reply:
Oh, it's not that `0`. Each `ClassName` will 256 bytes in file, e.g. 'motorway' will looks like `m o t o r w a y 0x00 0x00 0x00 ...`. Totally 8 `ClassName`s will in file. If only 5 valid values, then the last 3 `ClassName`s will all `0x00` of the 256 bytes.

Reply:
ok, max is 8.